gradle/javaToolchain.gradle

```diff
@@ -30,12 +30,56 @@ allprojects {
             }
         }
     }
+    project.pluginManager.withPlugin('org.graalvm.buildtools.native') {
+        def minRequiredGraalVMJavaVersion = 21
+        // By configuring the toolchains for the `org.graalvm.buildtools.native` plugin
+        // conditionally, we avoid Gradle errors caused by it failing to locate an installed GraalVM
+        // for Java SE older than 21. One situation when this is relevant is building from an IDE,
+        // where the `DEFAULT_JDK_VERSION` is likely used.
+        if (javaVersion >= minRequiredGraalVMJavaVersion) {
+            def javaLanguageVersion = JavaLanguageVersion.of(javaVersion)
+            // `JvmVendorSpec.GRAAL_VM` matches only GraalVM Community (https://github.com/graalvm/graalvm-ce-builds/releases),
+            // and does not match any other Graal distribution.
+            // That is, Gradle fails to locate any other installed distribution of GraalVM.
+            // Furthermore, there is no other way to express via the Gradle toolchain functionality
+            // that GraalVM must be used. The documentation of the `org.graalvm.buildtools.native` plugin
+            // says the following about this limitation:
+            // "be aware that the toolchain detection cannot distinguish between GraalVM JDKs
+            // and standard JDKs without Native Image support:
+            // if you have both installed on the machine, Gradle may randomly pick one or the other".
+            // Fortunately, `JvmVendorSpec.GRAAL_VM` makes things less hideous than that.
```

Comment:
This is a problem when running on Evergreen, because there Oracle GraalVM is installed. There are two possible ways forward:

1. Try using `GRAALVM_HOME` and see if it helps.
2. Request GraalVM Community to be installed on the computers that are used by Evergreen to run builds.

Reply:
> Try using GRAALVM_HOME and see if it helps.

I was unable to find a way to stop relying on the toolchain specification requiring `JvmVendorSpec.GRAAL_VM` even with `GRAALVM_HOME`. Reopened https://jira.mongodb.org/browse/DEVPROD-4928 and [requested](https://jira.mongodb.org/browse/DEVPROD-4928?focusedCommentId=6183735&focusedId=6183735&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-6183735) to install [GraalVM for JDK 21 Community 21.0.2](https://github.com/graalvm/graalvm-ce-builds/releases/tag/jdk-21.0.2) and [GraalVM for JDK 17 Community 17.0.9](https://github.com/graalvm/graalvm-ce-builds/releases/tag/jdk-17.0.9), though I don't have plans to use the latter.

Reply:
The current failure on Evergreen is exactly as I was saying (see the logs [here](https://parsley.mongodb.com/evergreen/mongo_java_driver_graalvm_native_image_app__version~7.0_os~linux_topology~replicaset_auth~noauth_ssl~nossl_jdk~jdk21_graalvm_native_image_app_patch_dc6c38b1d8ca730d1e45514ab1374bf4f78e7df6_65f4e523ee1f890007433076_24_03_16_00_17_40/0/task?bookmarks=0,1212&shareLine=897)):

> [2024/03/15 18:40:41.160] The GraalVM JDK distribution expected to be found at "/opt/java/jdk21-graal" by the Gradle toolchain functionality is
> [2024/03/15 18:40:44.589] java 21.0.2 2024-01-16 LTS
> [2024/03/15 18:40:44.589] Java(TM) SE Runtime Environment Oracle GraalVM 21.0.2+13.1 (build 21.0.2+13-LTS-jvmci-23.1-b30)
> [2024/03/15 18:40:44.589] Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 21.0.2+13.1 (build 21.0.2+13-LTS-jvmci-23.1-b30, mixed mode, sharing)
> ...
> [2024/03/15 18:41:55.246]    > No compatible toolchains found for request filter: {languageVersion=21, vendor=GRAAL_VM, implementation=vendor-specific} (auto-detect true, auto-download false)